File: lintcode/largest-number.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/largest-number/

class Solution:
    #@param num: A list of non negative integers
    #@return: A string
    def largestNumber(self, num):
        # write your code here
        numbers = []
        for n in num:
            numbers.append(str(n))
        numbers.sort(cmp = lambda x,y:cmp(x+y, y+x))
        res = ''.join(numbers[::-1])
        if res.count('0') == len(res):
            res = '0'
        return res

if __name__ == '__main__':
    s = Solution()
    print s. largestNumber([9,99,98])

File: leetcode/largest-number.cc
// Question: https://leetcode.com/problems/largest-number/

#include "headers.h"

struct cmp {
    bool operator() (const string& s1, const string& s2) {
        return s1+s2 > s2+s1;
    }
};

class Solution {
public:
    string largestNumber(vector<int>& nums) {
        if (nums.empty()) {
            return "";
        }

        vector<string> strs;
        for (auto n : nums) {
            stringstream ss;
            ss << n;
            strs.push_back(ss.str());
        }

        sort(strs.begin(), strs.end(), cmp());

        string ans;
        for (auto s : strs) {
            ans += s;
        }

        if (count(ans.begin(), ans.end(), '0') == ans.size()) {
            ans = '0';
        }
        return ans;
    }
};

int main(int argc, char** argv) {
    Solution s;
    vector<int> nums{0, 0};
    cout << s.largestNumber(nums) << endl;
    return 0;
}
